smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/traits/ContextIndex.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.rulesengine.traits;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.KnowledgeIndex;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.ToShapeId;
+import software.amazon.smithy.utils.SmithyUnstableApi;
+
+/**
+ * Resolves an indexes the context parameters in the model.
+ */
+@SmithyUnstableApi
+public final class ContextIndex implements KnowledgeIndex {
+    private final Map<ShapeId, Map<String, ClientContextParamDefinition>> serviceParams = new HashMap<>();
+    private final Map<ShapeId, Map<String, StaticContextParamDefinition>> operationParams = new HashMap<>();
+    private final Map<ShapeId, Map<MemberShape, ContextParamDefinition>> operationMemberParams = new HashMap<>();
+
+    public ContextIndex(Model model) {
+        for (ServiceShape serviceShape : model.getServiceShapes()) {
+            serviceShape.getTrait(ClientContextParamsTrait.class).ifPresent(trait -> {
+                serviceParams.put(serviceShape.toShapeId(), trait.getParameters());
+            });
+        }
+
+        for (OperationShape operation : model.getOperationShapes()) {
+            operation.getTrait(StaticContextParamsTrait.class).ifPresent(endpointParamsTrait -> {
+                operationParams.put(operation.toShapeId(), endpointParamsTrait.getParameters());
+            });
+            operation.getInput().ifPresent(shapeId -> {
+                StructureShape input = model.expectShape(shapeId, StructureShape.class);
+                Map<MemberShape, ContextParamDefinition> parameterReferenceMap = new LinkedHashMap<>();
+                input.members().forEach(memberShape -> {
+                    memberShape.getTrait(ContextParamTrait.class).ifPresent(contextParamTrait -> {
+                        parameterReferenceMap.put(memberShape, ContextParamDefinition.builder()
```

Comment:
I think this is a kind of deficiency with the current design that separates traits from the data they contain